### PR TITLE
Remove boost dependency

### PR DESCRIPTION
**Pull request by kylealanhale**
The primary purpose of this pull request is to remove the boost regex dependency and use C++11’s regex support instead, as well as providing some custom trim functions to replace boost’s.

Additionally, some refactoring and logic work were done to handle edge cases, especially in the way that multi-line sections and newlines are treated that differ from the ruby implementation. Tests were added for these scenarios.

The travis config was updated to run jobs for both GCC 4.9 and Clang, and both are passing, although GCC’s implementation slightly differed, as you can see in the last commit.

A new Xcode project was also added, with unit tests wired up and passing.

### CODE REVIEW

**Comment by mrtazz** [comment]
Oh wow, this is great! Thank you so much for taking the time to contribute! It will take me some time to look it over but I'm really happy about this.

**Comment by shiva** [comment]
@kylealanhale Didn't notice that you had already raised these for review! Awesome.

@mrtazz Nice that you have a Travis CI build going for this. It would be great when these changes go in, and I can abandon my fork!

**Comment by shiva** [comment]
@mrtazz Wondering if you have had a chance to consider merging this patch? I'm curious about some of the other branches, and support for CMake based builds, and some other features like "the fix for partial file load" etc.

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,36 @@
...
+      - gcc-4.9
+      - g++-4.9
+      - clang
+before_install:
+  - if [ "$CXX" = "g++" ]; then 
+      export CXX="g++-4.9" CC="gcc-4.9";
+      export CXXFLAGS="-std=c++11";
+    elif [ "$CXX" = "clang++" ]; then
+      export CXXFLAGS="-std=c++11 -stdlib=libc++ -DGTEST_USE_OWN_TR1_TUPLE=1";
+    fi
```
Comment (mrtazz) [inline .travis.yml]: can we move those into the Makefile and the gtest build utils script? Or is this a specific travis problem? I'd much rather have it close to the actual build if this will be a problem outside of Travis.

File: .travis.yml [hunk 2]
```diff
@@ -1,6 +1,36 @@
...
+      export CXXFLAGS="$CXXFLAGS -L`pwd`/libcxx/lib -I`pwd`/libcxx/include";
+      export LD_LIBRARY_PATH="`pwd`/libcxx/lib:$LD_LIBRARY_PATH";
+    fi
 before_script:
-  - sudo apt-get install libgtest-dev libboost-regex-dev build-essential
   - ./utils/build_gtest.sh
+  - autoreconf -i && ./configure && make check
+script: ./test-program
+notifications:
+  email: false
```
Comment (mrtazz) [inline .travis.yml]: we can remove the notifications thing here. I'm fine with the defaults. Or was there a specific reason why this was added?

File: src/template.cpp [hunk 3]
```diff
@@ -7,14 +7,35 @@
...
+#include <algorithm>
 
 #include <plustache/plustache_types.hpp>
 #include <plustache/context.hpp>
 
 using namespace Plustache;
 
+
+// trim from start
+static inline std::string &ltrim(std::string &s)
```
Comment (mrtazz) [inline src/template.cpp]: can you add doxygen style comments here (similar to what the existing functions have) and explain the code a little as well. This single magic line is pretty cryptic to read.

File: src/template.cpp [hunk 4]
```diff
@@ -7,14 +7,35 @@
...
+
+// trim from start
+static inline std::string &ltrim(std::string &s) 
+{
+    s.erase(s.begin(), std::find_if(s.begin(), s.end(), std::not1(std::ptr_fun<int, int>(std::isspace))));
+    return s;
+}
+
+// trim from end
+static inline std::string &rtrim(std::string &s)
```
Comment (mrtazz) [inline src/template.cpp]: same here. please add doxygen style headers and explain the code a bit.

File: src/template.cpp [hunk 5]
```diff
@@ -7,14 +7,35 @@
...
+
+// trim from end
+static inline std::string &rtrim(std::string &s) 
+{
+    s.erase(std::find_if(s.rbegin(), s.rend(), std::not1(std::ptr_fun<int, int>(std::isspace))).base(), s.end());
+    return s;
+}
+
+// trim from both ends
+static inline std::string &trim(std::string &s)
```
Comment (mrtazz) [inline src/template.cpp]: doxygen headers please.

File: src/template.cpp [hunk 6]
```diff
@@ -80,23 +109,23 @@ std::string template_t::render_tags(const std::string& tmplate,
...
     std::string rest = "";
     std::string::const_iterator start, end;
-    boost::match_results<std::string::const_iterator> matches;
+    std::match_results<std::string::const_iterator> matches;
     start = tmplate.begin();
     end = tmplate.end();
     // return whole std::string when no tags are found
     if (!regex_search(start, end, matches, tag,
-                      boost::match_default | boost::format_all))
+                      std::regex_constants::match_default | std::regex_constants::format_default))
```
Comment (mrtazz) [inline src/template.cpp]: can you newline and indent this to fit 80 chars please?

File: src/template.cpp [hunk 7]
```diff
@@ -80,23 +109,23 @@ std::string template_t::render_tags(const std::string& tmplate,
...
     {
         ret = tmplate;
     }
     // loop through tags and replace
     while (regex_search(start, end, matches, tag,
-                        boost::match_default | boost::format_all))
+                        std::regex_constants::match_default | std::regex_constants::format_default))
     {
```
Comment (mrtazz) [inline src/template.cpp]: also newline and indent here

File: src/template.cpp [hunk 8]
```diff
@@ -141,17 +170,17 @@ std::string template_t::render_tags(const std::string& tmplate,
 
         // replace
         ret += regex_replace(text, tag, repl,
-                             boost::match_default | boost::format_all);
+                             std::regex_constants::match_default | std::regex_constants::format_default);
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

File: src/template.cpp [hunk 9]
```diff
@@ -141,17 +170,17 @@ std::string template_t::render_tags(const std::string& tmplate,
...
-          boost::regex delim("(.+?) (.+?)=");
+          std::regex delim("(.+?) (.+?)=");
           // match object
-          boost::match_results<std::string::const_iterator> delim_m;
+          std::match_results<std::string::const_iterator> delim_m;
           // search for the delimiters
-          boost::regex_search(matches[2].first, matches[2].second, delim_m, delim,
-                              boost::match_default | boost::format_all);
+          std::regex_search(matches[2].first, matches[2].second, delim_m, delim,
+                              std::regex_constants::match_default | std::regex_constants::format_default);
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent for 80 chars

File: src/template.cpp [hunk 10]
```diff
@@ -182,30 +214,32 @@ std::string template_t::render_sections(const std::string& tmplate,
...
     std::string::const_iterator start, end;
-    boost::match_results<std::string::const_iterator> matches;
+    std::match_results<std::string::const_iterator> matches;
     start = tmplate.begin();
     end = tmplate.end();
     // return the whole template if no sections are found
-    if (!boost::regex_search(start, end, matches, section,
-                             boost::match_default | boost::format_all))
+    if (!std::regex_search(start, end, matches, section,
+                             std::regex_constants::match_default | std::regex_constants::format_default))
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

File: src/template.cpp [hunk 11]
```diff
@@ -182,30 +214,32 @@ std::string template_t::render_sections(const std::string& tmplate,
...
     {
         ret = tmplate;
     }
     // loop through sections and render
-    while (boost::regex_search(start, end, matches, section,
-                               boost::match_default | boost::format_all))
+    while (std::regex_search(start, end, matches, section,
+                               std::regex_constants::match_default | std::regex_constants::format_default))
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

File: src/template.cpp [hunk 12]
```diff
@@ -225,37 +259,53 @@ std::string template_t::render_sections(const std::string& tmplate,
...
+                repl += template_t::render_tags(content, small_ctx);
             }
         }
+        // Hide section with empty content
         else repl.assign("");
-        ret += boost::regex_replace(text, section, repl,
-                                    boost::match_default | boost::format_all);
-        rest.assign(matches[0].second, end);
-        start = matches[0].second;
+
```
Comment (mrtazz) [inline src/template.cpp]: can you please wrap this in curly braces as well. I know it's not required but I think it makes the code clearer.

File: src/template.cpp [hunk 13]
```diff
@@ -225,37 +259,53 @@ std::string template_t::render_sections(const std::string& tmplate,
...
         }
+        // Hide section with empty content
         else repl.assign("");
-        ret += boost::regex_replace(text, section, repl,
-                                    boost::match_default | boost::format_all);
-        rest.assign(matches[0].second, end);
-        start = matches[0].second;
+        
+        // Replace matched section with generated text
+        ret += std::regex_replace(text, section, repl, std::regex_constants::match_default | std::regex_constants::format_default);
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

File: src/template.cpp [hunk 14]
```diff
@@ -316,26 +366,26 @@ std::string template_t::html_escape(const std::string& s)
...
     std::string::const_iterator start, end;
-    boost::match_results<std::string::const_iterator> matches;
+    std::match_results<std::string::const_iterator> matches;
     start = s.begin();
     end = s.end();
     // return original std::string if nothing is found
-    if (!boost::regex_search(start, end, matches, escape_chars,
-                             boost::match_default | boost::format_all))
+    if (!std::regex_search(start, end, matches, escape_chars,
+                             std::regex_constants::match_default | std::regex_constants::format_default))
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

File: src/template.cpp [hunk 15]
```diff
@@ -316,26 +366,26 @@ std::string template_t::html_escape(const std::string& s)
...
     {
         ret = s;
     }
     // search for html chars
-    while (boost::regex_search(start, end, matches, escape_chars,
-                               boost::match_default | boost::format_all))
+    while (std::regex_search(start, end, matches, escape_chars,
+                               std::regex_constants::match_default | std::regex_constants::format_default))
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

File: src/template.cpp [hunk 16]
```diff
@@ -316,26 +366,26 @@ std::string template_t::html_escape(const std::string& s)
...
         std::string key(matches[0].first, matches[0].second);
         std::string text(start, matches[0].second);
-        boost::algorithm::trim(key);
+        trim(key);
         std::string repl;
         repl = escape_lut[key];
-        ret += boost::regex_replace(text, escape_chars, repl,
-                                    boost::match_default | boost::format_all);
+        ret += std::regex_replace(text, escape_chars, repl,
+                                    std::regex_constants::match_default | std::regex_constants::format_default);
```
Comment (mrtazz) [inline src/template.cpp]: newline and indent to fit in 80 chars.

**Comment by mrtazz** [comment]
@kylealanhale I left some comments there. Most of it was really just stylistic as I think the code overall looks good. It would be great if you could in general try to make it break lines at 80 chars and remove trailing whitespace as it makes it easier for the next person to contribute regardless of editor/IDE. I also realized that this could all be made explicit in a `CONTRIBUTING.md` so I'll make sure to add one.

**Comment by shiva** [comment]
@kylealanhale Would you prefer that I work on the comments provided @mrtazz ? Let me know if you are otherwise pre-occupied, and I will find some time for these changes.

**Comment by kylealanhale** [comment]
@shiva I should be able to knock these out in the next couple of days.

File: .travis.yml (hunk 1 above)
Comment (kylealanhale) [inline .travis.yml]: These are for the sake of Travis, although they're good clues on dependencies needed for building anywhere. However, I believe you have the `-DGTEST_USE_OWN_TR1_TUPLE=1` documented already for clang, and I think I added something about gcc-4.9 and c++11, but I'll double check and make sure it's all there somewhere.

File: .travis.yml (hunk 2 above)
Comment (kylealanhale) [inline .travis.yml]: I'll nix this; I had to push a travis build over and over to get it to work right, so I turned them off and forget to get rid of it.

**Comment by kylealanhale** [comment]
Took care of those; let me know if you see anything else. Otherwise, I think this is ready to go.

**Comment by kylealanhale** [comment]
What's your versioning approach? Since this is a substantial change that will break for people who for any reason can't compile with c++11 support, it might warrant a new release. And maybe a 0.3.x release before merging this to catch any recent changes, which might be useful for anyone that wants to use the old Boost version.

**Comment by mrtazz** [comment]
@kylealanhale yea that's a good point. I wanted to make it a new release anyways. I'll check what we have in master since the last release and will tag one before merging. Thanks again for doing all this work!
